### PR TITLE
Add semantic conventions for cloud provider-specific resource attributes

### LINKED ISSUE #1098: Provide specification for cloud-provider-specific resource attributes
**What are you trying to achieve?**

The current resource specs define a broad range of useful attributes that exist for certain types of compute infrastructure regardless of the who provides it, like `container.id` or `faas.name`. However there are many other attributes that OpenTelemetry users would appreciate having in their spans which are specific to the cloud/infrastructure provider. For example, AWS Elastic Container Service (ECS) has metadata related to its "tasks" and "launch type" that do not fit neatly into the existing generic attributes.

In order to record these resource attributes in a consistent way across SDKs and the collector, I think their semantic conventions should live here in the [OTel spec for resources](https://github.com/willarmiros/opentelemetry-specification/tree/master/specification/resource/semantic_conventions) under a subdirectory like `cloud_provider` indicating these attributes are only present in a specified provider's resource. For example, `cloud_provider/azure` could host specific attributes about Azure Functions, `cloud_provider/aws` could host specific attributes about ECS, etc.

To avoid bloat, we can constrain the core spec to only contain provider-specific resources from the supported `cloud_provider`s, which are currently `aws`, `gcp`, and `azure`.

**Additional context.**

This is related to #968, except this issue is about vendor-specific resource semantic convention as opposed to those for instrumentation. I will be following up with a PR that proposes a solution shortly. Always open to feedback here and on the PR when open.

### PR DESCRIPTION
**Pull request by willarmiros**
Fixes #1098

## Changes

This is an implementation of my proposal in #1098 to include a series of resource attributes that are specific to AWS, and not useful as generalized attributes. I believe this proposal is extendable to include attributes from other cloud providers like GCP, Azure, and any other providers involved in OpenTelemetry. I'd appreciate feedback on anything from the location of this within the spec to the naming of attributes.

One thing syntactically I wasn't sure of was how to describe attributes with common prefixes within a group. For example, for `aws.ecs.task.arn` and `aws.ecs.task.family`, should I create a new group called `aws.ecs.task` with 2 attributes (`arn` and `family`), or should I simply include `task.arn` and `task.family` as attributes within the larger `aws.ecs` group, as I've done below?

I've also included a new attribute on the `cloud` group called `infrastructure.service`. Please see the modified `cloud.md` for more details. I think that such an attribute would be useful for users who want to quickly identify the type of compute platform a trace/metric is originating from. I've used the syntax that we use in AWS X-Ray, but I am open to modifying it (e.g. removing the cloud provider prefix since that is already known from `cloud.provider` attribute, or changing the delimiter). Or I can move the attribute to the AWS spec if others feel it doesn't belong in `cloud.md`. 

Related issue: #968

### CODE REVIEW

File: semantic_conventions/resource/cloud.yaml [hunk 1]
```diff
@@ -13,7 +13,7 @@ groups:
               brief: 'Amazon Web Services'
             - id: Azure
               value: 'azure'
-              brief: 'Amazon Web Services'
+              brief: 'Microsoft Azure'
```
Comment (anuraaga) [inline semantic_conventions/resource/cloud.yaml:16]: 😄

File: semantic_conventions/resource/cloud.yaml [hunk 2]
```diff
@@ -37,3 +37,12 @@ groups:
         note: >
           In AWS, this is called availability-zone.
         examples: ['us-central1-a']
+      - id: infrastructure.service
```
Comment (anuraaga) [inline semantic_conventions/resource/cloud.yaml:40]: It's probably good to separate this into a separate PR since it's broader reaching than AWS-specific conventions.

File: semantic_conventions/resource/cloud.yaml [hunk 3]
```diff
@@ -37,3 +37,12 @@ groups:
...
+        type: string
+        brief: >
+          Description of cloud resource in use.
+        note: >
+          The first entry should generally be a cloud provider, followed by a product
+          category, then finally a particular piece of compute infrastructure. Each entry
+          is delimited by a double colon (::).
+        examples: ['AWS::EC2::Instance', 'Azure::Compute::VM', 'GCP::ComputeEngine::VM']
```
Comment (anuraaga) [inline semantic_conventions/resource/cloud.yaml:48]: AWS, etc are already in `cloud.provider` so this should only need something like `ec2`, `aks`, `cloud run`. Of course we'll want insight from a GCP and Azure member for some examples :)

X-Ray uses these extended names but it's the X-Ray exporter that can format the string correctly but not what we'd require in the semantic convention which should just be a precise semantic string.

File: semantic_conventions/resource/cloud_provider/aws/aws.yaml [hunk 4]
```diff
@@ -0,0 +1,34 @@
+groups:
```
Comment (thisthat) [inline semantic_conventions/resource/cloud_provider/aws/aws.yaml:1]: This seems more related to logging than AWS in the general sense.

File: specification/resource/semantic_conventions/README.md [hunk 5]
```diff
@@ -135,3 +135,9 @@ Attributes defining a running environment (e.g. Operating System, Cloud, Data Ce
 Version attributes, such as `service.version`, are values of type `string`. They are
 the exact version used to identify an artifact. This may be a semantic version, e.g., `1.2.3`, git hash, e.g.,
 `8ae73a`, or an arbitrary version string, e.g., `0.1.2.20210101`, whatever was used when building the artifact.
+
+## Provider-Specific Attributes
+
+Attributes that are only applicable to resources from a specific cloud provider. Currently, these
+resources can only be defined for providers listed as a valid `cloud.provider` in
+[Cloud](./cloud.md). Provider-specific attributes all reside in the `cloud_provider` directory.
```
Comment (thisthat) [inline specification/resource/semantic_conventions/README.md:143]: Maybe add the list of providers for which we have some semantic convention here

**Review by thisthat — COMMENTED** [review]
Thank you for your contribution :)
I would suggest adding a README.md file inside the `aws` folder that contains the list of AWS related semantic conventions, similar to what we have for [trace semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/tree/master/specification/trace/semantic_conventions)  :)

File: semantic_conventions/resource/cloud.yaml (hunk 2 above)
Comment (willarmiros) [reply semantic_conventions/resource/cloud.yaml:40]: Ok will separate it into a new PR with your feedback included.

File: semantic_conventions/resource/cloud.yaml (hunk 3 above)
Comment (willarmiros) [reply semantic_conventions/resource/cloud.yaml:48]: Makes sense, I sorta figured as much. I'll get rid of the delimiters and just specify the compute platform.

File: specification/resource/semantic_conventions/README.md (hunk 5 above)
Comment (willarmiros) [reply specification/resource/semantic_conventions/README.md:143]: Sure, I'll include a list of valid cloud providers more explicitly.

File: semantic_conventions/resource/cloud_provider/aws/aws.yaml (hunk 4 above)
Comment (willarmiros) [reply semantic_conventions/resource/cloud_provider/aws/aws.yaml:1]: Would you prefer I move this to somewhere in the [logs spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/logs/overview.md)? The `loggroup` and `logstream` fields are specific terms to AWS CloudWatch, but if there are more generic terms that would apply to other logging backends but still convey the same meaning, I'd be happy to switch the wording.

Comment (thisthat) [reply semantic_conventions/resource/cloud_provider/aws/aws.yaml:1]: I am not familiar with logs, but I guess there will be semantic conventions for it as well!
For the moment, I would suggest changing the id to `aws.log` to avoid a naming clash with other more general `aws` semantic conventions.

**Comment by willarmiros** [comment]
Made an additional change to make the AWS log attributes to be string arrays. As I noted in the spec:

> Multiple log groups must be supported for cases like multi-container applications, where a single application has sidecar containers, and each write to their own log group.

**Comment by willarmiros** [comment]
@thisthat I've applied your suggestions, please give a second look when available :)

File: specification/resource/semantic_conventions/README.md [hunk 6]
```diff
@@ -135,3 +135,14 @@ Attributes defining a running environment (e.g. Operating System, Cloud, Data Ce
...
 `8ae73a`, or an arbitrary version string, e.g., `0.1.2.20210101`, whatever was used when building the artifact.
+
+## Cloud-Provider-Specific Attributes
+
+Attributes that are only applicable to resources from a specific cloud provider. Currently, these
+resources can only be defined for providers listed as a valid `cloud.provider` in
+[Cloud](./cloud.md) and below. Provider-specific attributes all reside in the `cloud_provider` directory.
+Valid cloud providers are:
+
+- [Amazon Web Services](https://aws.amazon.com/) (`aws`)
```
Comment (thisthat) [inline specification/resource/semantic_conventions/README.md:146]:
```suggestion
- [Amazon Web Services](https://aws.amazon.com/) ([`aws`](cloud_provider/aws/README.md))
```

**Review by thisthat — APPROVED** [review]
Thank you for your contribution :)

**Comment by anuraaga** [comment]
> Multiple log groups must be supported for cases like multi-container applications, where a single application has sidecar containers, and each write to their own log group.

@willarmiros is this true for telemetry though? Won't any single app process (Resource) only have a single log group since each sidecar container is a different process? I might be confused on it but I think it's appropriate for these to be singular, any metric resource can only be reporting to a single log group I think.

**Comment by willarmiros** [comment]
@anuraaga you're right that each container or process has its own log group, and I suppose I worded it as a feature even though it's more of a limitation. For ECS, the problem is that when calling the Task Metadata Endpoint from the collector container, we have no way of identifying which container is the application container. We just get a list of all containers defined for a task, and we get the container the collector is running on. So we can confidently exclude the collector's log group from the list of log groups, but if there are other sidecar containers we can't distinguish them from the main app container.

Similarly, in EC2, we get the log group data from the [agent config file](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/AgentReference.html#agent-configuration-file), which can have any number of log groups. Again, we can't distinguish which one is the "right" one. This is why log groups on X-Ray segments are also modeled as an array, even though an X-Ray segment is also only supposed to correspond to one resource.

The only workaround us we get some sort of user input, e.g. in `config.yaml`. However detector-specific config (e.g. configuration for just the ECS resource detector) doesn't exist and even if we added it, it would still be cumbersome to users IMO. At that point, we might as well just have the user input their log group themselves :)

**Comment by willarmiros** [comment]
@anuraaga when you get a chance a second review would be appreciated :)

**Review by anuraaga — APPROVED** [review]
Thanks!

**Comment by anuraaga** [comment]
@open-telemetry/specs-approvers Can anyone take a look at this PR? Thanks!

**Comment by tigrannajaryan** [comment]
> @open-telemetry/specs-approvers Can anyone take a look at this PR? Thanks!

@anuraaga I believe we need to resolve https://github.com/open-telemetry/opentelemetry-specification/issues/1151 which blocks this.

**Comment by willarmiros** [comment]
@tigrannajaryan I plan on attending the spec issue scrub tomorrow morning to discuss :) if there's a better time/place to raise please let me know

**Comment by tigrannajaryan** [comment]
> @tigrannajaryan I plan on attending the spec issue scrub tomorrow morning to discuss :) if there's a better time/place to raise this please let me know

The spec meeting on Tue may be a better time. Also it would be useful to invite some attention to this issue so that it is discussed offline too. @open-telemetry/specs-approvers please chime in at https://github.com/open-telemetry/opentelemetry-specification/issues/1151 if you have thoughts.

**Comment by willarmiros** [comment]
Sounds good. Also to be clear, #1151 blocks #1112 but I don't think it should block this PR. This PR is strictly adding AWS-specific resource fields and adding a place for other cloud-provider-specific attributes to live.

#1112 involves adding a new generic `cloud` attribute, so understandably in need of more discussion. Regardless I'll plan on attending Tuesday to chat!

File: specification/resource/semantic_conventions/cloud_provider/aws/ecs.md [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+**type:** `aws.ecs`
+
+**Description:** Resources used by AWS Elastic Container Service (ECS).
+
+<!-- semconv aws.ecs -->
+| Attribute  | Type | Description  | Example  | Required |
+|---|---|---|---|---|
+| `aws.ecs.container.arn` | string | The Amazon Resource Name (ARN) of an [ECS container instance](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ECS_instances.html). | `arn:aws:ecs:us-west-1:123456789123:container/32624152-9086-4f0e-acae-1a75b14fe4d9` | No |
+| `aws.ecs.cluster.arn` | string | The ARN of an [ECS cluster](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/clusters.html). | `arn:aws:ecs:us-west-2:123456789123:cluster/my-cluster` | No |
```
Comment (hossain-rayhan) [inline specification/resource/semantic_conventions/cloud_provider/aws/ecs.md:11]: When we get task metadata from ECS Task Metadata Endpoint, it gives us the ClusterName for EC2 Launch Type. Can't we add one for `aws.ecs.cluster.name`?

Comment (willarmiros) [reply specification/resource/semantic_conventions/cloud_provider/aws/ecs.md:11]: I originally had it as `cluster.name` but, since I was populating the ARN for nearly all other fields and consumers could easily derive cluster name from the ARN, I figured recording the cluster ARN made more sense. On Fargate we are provided the cluster ARN by default, and on EC2 launch type, we have the region and account ID from the task ARN, so you can always construct the cluster ARN like I do [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/processor/resourcedetectionprocessor/internal/aws/ecs/ecs.go#L83).

All this being said, I can still add a new field for the cluster name. I just wanted to reduce the possibility for redundant info.

Comment (hossain-rayhan) [reply specification/resource/semantic_conventions/cloud_provider/aws/ecs.md:11]: I guess  very few customers care about the whole ARN. Most useful info here is the resource name (`ClusterName` or `TaskID`). It's also implies that all the exporters should write their logic to extract resource name from ARN. It's not only extracting the resource name, they need to do other checks before that. If resource attribute map does not have the name fields, extract them and apply. The logic is not super complex but extra work and check. 
So, in a sense, we are blocking our customers if any specific exporter doesn't come with these logics. 

Also, I am kind of convinced with the idea of less redundant info. I am OK to follow the experts opinion.

Comment (willarmiros) [reply specification/resource/semantic_conventions/cloud_provider/aws/ecs.md:11]: As discussed offline: We're ok with this as is for now, and in the future if customers ask for it we can add cluster name, task ID, etc to this spec.

**Review by reyang — APPROVED** [review]
LGTM.
